File: code/C++/841-Keys and Rooms.cpp
// There are n rooms labeled from 0 to n - 1 and all the rooms are locked except for room 0. Your goal is to visit all the rooms. However, you cannot enter a locked room without having its key.

// When you visit a room, you may find a set of distinct keys in it. Each key has a number on it, denoting which room it unlocks, and you can take all of them with you to unlock the other rooms.

// Given an array rooms where rooms[i] is the set of keys that you can obtain if you visited room i, return true if you can visit all the rooms, or false otherwise.

 

// Example 1:

// Input: rooms = [[1],[2],[3],[]]
// Output: true
// Explanation: 
// We visit room 0 and pick up key 1.
// We then visit room 1 and pick up key 2.
// We then visit room 2 and pick up key 3.
// We then visit room 3.
// Since we were able to visit every room, we return true.
// Example 2:

// Input: rooms = [[1,3],[3,0,1],[2],[0]]
// Output: false
// Explanation: We can not enter room number 2 since the only key that unlocks it is in that room.
 

// Constraints:

// n == rooms.length
// 2 <= n <= 1000
// 0 <= rooms[i].length <= 1000
// 1 <= sum(rooms[i].length) <= 3000
// 0 <= rooms[i][j] < n
// All the values of rooms[i] are unique.

class Solution {
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms) {
        int n = rooms.size();
        vector<int> seen(n);
        seen[0] = 1;
        priority_queue<int> que;
        for (int key: rooms[0]) {
            que.emplace(key);
        }
        while (!que.empty()) {
            int key = que.top();
            que.pop();
            if (!seen[key]) {
                seen[key] = 1;
                for (int key_ : rooms[key]) {
                    que.emplace(key_);
                }
            }
        }
        for (int i = 0; i < n; i ++) {
            if (!seen[i]) {
                return false;
            }
        }
        return true;
    }
};

File: code/841-Keys and Rooms/841-Keys and Rooms.py
# -*- coding: utf-8 -*
"""
@author: Kai Chen
@file: 841-Keys and Rooms.py
@time: 2020/8/31 9:54
@desc:
There are N rooms and you start in room 0.
Each room has a distinct number in 0, 1, 2, ..., N-1, and each room may have some keys to access the next room.

Formally, each room i has a list of keys rooms[i],
and each key rooms[i][j] is an integer in [0, 1, ..., N-1] where N = rooms.length.
A key rooms[i][j] = v opens the room with number v.

Initially, all the rooms start locked (except for room 0).

You can walk back and forth between rooms freely.

Return true if and only if you can enter every room.

Example 1:

Input: [[1],[2],[3],[]]
Output: true
Explanation:
We start in room 0, and pick up key 1.
We then go to room 1, and pick up key 2.
We then go to room 2, and pick up key 3.
We then go to room 3.  Since we were able to go to every room, we return true.
Example 2:

Input: [[1,3],[3,0,1],[2],[0]]
Output: false
Explanation: We can't enter the room with number 2.
Note:

1 <= rooms.length <= 1000
0 <= rooms[i].length <= 1000
The number of keys in all rooms combined is at most 3000.
Accepted

"""

from collections import deque

# bfs
# class Solution(object):
#     def canVisitAllRooms(self, rooms):
#         """
#         :type rooms: List[List[int]]
#         :rtype: bool
#         """
#         n = len(rooms)
#         visited = [0 for _ in range(n)]
#         visited[0] = 1
#         stack = deque([0])
#         while stack:
#             room_idx = stack.popleft()
#             visited[room_idx] = 1
#             for key in rooms[room_idx]:
#                 if not visited[key]:
#                     stack.append(key)
#         return False if 0 in visited else True


# dfs
class Solution(object):
    def canVisitAllRooms(self, rooms):
        """
        :type rooms: List[List[int]]
        :rtype: bool
        """
        n = len(rooms)
        visited = set([0])

        def dfs(room):
            visited.add(room)
            for key in rooms[room]:
                if key not in visited:
                    dfs(key)

        dfs(0)
        return len(visited) == n

            

if __name__ == '__main__':
    s = Solution()
    ipts = [
        [[1], [2], [3], []],
        [[2], [], [1]],
        [[1, 3], [3, 0, 1], [2], [0]]
    ]
    for ipt in ipts:
        print(s.canVisitAllRooms(ipt))
